### ISSUE TITLE
Cannot figure out how to do "sensible" logging in cloud functions

### ISSUE THREAD

**Issue by npomfret** [googleapis/nodejs-logging#633]
I want to be able to do the following (but have read docs and tried everything I can think of but cannot work it out).  Any advice would be appreciated.

... in a cloud-function (http and pubsub)

 * log a message with a stacktrace (*on a single line*)
 * log json objects *on a single line*
 * log at specified log levels (error, warning, info etc)
 * log errors that cause the function to crash (so that I can see the error!)

https://stackoverflow.com/questions/58132635/better-logging-from-pub-sub-cloud-function-in-gcp

I can get some of these things working in isolation, but I cannot find a solution that covers all of them.

**Comment by bcoe** [comment]
@npomfret have you tried the `@google-cloud/logging-bunyan` or `@google-cloud/logging-winston` abstractions. I think, as that stack trace indicates, they can provide some of the niceties you're looking for.

If we are seeing bugs in `@google-cloud/logging-winston` or `bunyan` that makes them difficult to use in cloud functions, it would be good for us to address those issues (since both these libraries are built on `@google-cloud/logging`, I think similar issues would arise using it directly).

Long story short, I would suggest trying one of these abstractions (`bunyan` or `winston`) and if we're seeing bugs, we can work on fixing them :+1:

**Comment by npomfret** [comment]
Yep, tried logging-bunyan which sort of works, but doesn't log anything when the function crashes (which is typically when I want it most).

The problem is that it's ansychonous, and when a pubsub function finishes it looks like any logging after the fact is inhibited.  The same is not true for http functions, Bunyan works quite well in that situation.

**Comment by bcoe** [comment]
@npomfret I've directed this question to one of my peers who works more closely with cloud run.

One thought that does come to mind for me, you could use the underlying logger and await the log:

```
  // Writes the log entry
  await log.write(entry);
```

the downside to this, is you're waiting potentially `100ms` for the log entry top be sent, which can slow down your application.

**Comment by grayside** [comment]
Cloud Functions does not support structured logging via `console.log`, and will set severity to either `error` or `info` based on whether stderr or stdout are used.

Furthermore, CPU is withdrawn between function invocations and as a result logs that were meant to be flushed can end up stuck until a future invocation of the function hits the same instance... assuming that instance isn't cleaned up as part of scale-down.

The bunyan-logging and winston-logging libraries use the API surface of the logging library, but under the hood use the Cloud Logging client library, including it's buffering mechanisms, to deliver the logs to Stackdriver.

Cloud Run does support structured logging the way you want to approach it, and if these are HTTP functions the use of https://github.com/GoogleCloudPlatform/functions-framework-nodejs would facilitate migrating to Cloud Run.

**Comment by npomfret** [comment]
@grayside, thanks for the info.

> Cloud Functions does not support structured logging via console.log, and will set severity to either error or info based on whether stderr or stdout are used.

This can't be right, look at the first of my two images above - it clearly shows console logging working perfectly at different log levels.  This is output from an http cloud function.  The same is NOT true in a pubsub function, console logging levels work as you've described.

My experience of bunyan-logging in a pubsub function is that it fails to log if the function execution results in an exception.  You typically don't get any logs in this case and that isn't a viable solution for me.

What I want, is a solution that works in all cloud functions (well http and pubsub!) that allows me to

 * log at different levels (debug, info etc)
 * log exceptions on a single line
 * log json objects on a single line (or to a jsonPayload field)
 * log successfully when the function's code crashes

It sounds like this just isn't possible, am I right?

**Comment by bcoe** [comment]
@grayside how does the structured logging work in Cloud Run, perhaps you're right that this is a better choice for @npomfret.

@npomfret this product is pretty darn similar, you just run an arbitrary docker container, rather than a Node.js cloud function (you can still run Node.js as the container).

**Comment by bcoe** [comment]
@npomfret for the PubSub use case, I'd be tempted to potentially `await` the log write. We could potentially add a configuration setting to bunyan and winston that allowed you to enable `await`, rather than buffered writes (just a thought, perhaps it would be too much of a performance hit).

**Comment by merlinnot** [comment]
@bcoe I tried that in the past. Without batching of logs one would burst through quotas of StackDriver incredibly quickly. I'd advise against this solution.

**Comment by npomfret** [comment]
I can't be the only person who uses cloud functions and wants this fairly standard logging functionality.  Has this really not come up before?

**Comment by npomfret** [comment]
Perhaps the solution is to use Bunyan for everything **except** for when you need to log the error that caused the function to crash, and use `console.error` in that situation.

Or, are we saying the Buynan will not always work in cloud functions due to it's asynchronous nature (some logs will get lost).

**Comment by bcoe** [comment]
@npomfret you're hitting the nail on the head that `uncaughtException` and `unhandledRejection` are the potential problems, if the process is about to exit, you can't perform asynchronous operations. Having said this, you should be able to `console.error` in these exceptional cases.

What I might recommend, is trying to make your application as resilient as possible, such that you have a top level `catch` and an unhandled error is unlikely, in the main body of your program you could then use `bunyan`.

If there is additional logging you would like to perform in an exceptional case, you could listen to `on('unhandledRejection')` and `on('uncaughtException')` and add additional logging with `console.error()`.

**Comment by npomfret** [comment]
I'm not sure this is correct.  

I'm not talking about unhandledRejection or uncaughtException errors.  The situation I observed was that when an error occurred in my function, I would catch, log and re-throw it (so that the message gets redelivered at a later date).  In this situation I would see no logs at all using bunyan.

**Comment by npomfret** [comment]
Does anyone have a solution for this yet?  I still can't get reliable & sensible logging in cloud functions

**Comment by bcoe** [comment]
@npomfret I think some of the issues being bumped into here are currently limitations of the runtime environment itself, as @grayside points out.

There have been some discussions about improvements that could be made to function execution to make this experience better.

It might be worth adding a comment [here](https://issuetracker.google.com/savedsearches/559764), on the issue tracker. This gives you a more direct line to the product team.

**Comment by ernsheong** [comment]
> console.method also doesn't log at the correct level in pubsub functions (everything comes out as debug)

This is what I am seeing as well. When the function is a background function, `console.method` no longer works as advertised.

Any fix for this?

**Comment by npomfret** [comment]
I think I've solved it.  The key (as advised) is to make sure you wait for all logs to have been accepted by the `@google-cloud/logging` framework.  

Take a look at this: https://gist.github.com/npomfret/deef51b186e915504a7c77966c88a4f9

Usage:

```
const logger = createLogger('cloudfunctions.googleapis.com/cloud-functions', PROJECT_ID);
logger.divertConsoleToLogger();// this allows you to use console.xxx() methods
// or global.logger = logger;

const app = async (pubSubEvent, context) => {
    try { 
        const {timestamp, eventId} = context;
 
        logger.labels = {
          execution_id: eventId, // this very useful in stackdriver 
          latency: Date.now() - Date.parse(timestamp)
        };

        await /* your event handling code goes here */;

        await logger.flush();
        logger.labels = undefined;
    } catch (e) {
        await logger.flush();
        logger.labels = undefined;

        error(`${context.eventId} function ${functionName} crashed`, e);

        throw e;
    }
};

module.exports = {
    app
};
```

In a pubsub function (as described above), you need to make sure you catch any exceptions thrown by your function, and before you re-throw them, make sure that you call `await logger.flush();` which will return a promise that resolves only when all logs have been processed.

Http functions are similar but a little more complicated to describe with an example because it depends how the flow is handled by your middleware. Essentially the theory is the same, no matter how your functions ends (in an exception or not), you need to make sure to call `await logger.flush();`.

**Comment by npomfret** [comment]
Sorry, ignore that.  No matter what I do I cannot get `@google-cloud/logging` to work reliably in pubsub cloud functions.

**Comment by npomfret** [comment]
It's worth mentioning too, that using `@google-cloud/logging` with any sort of load results in the cloud functions crashing (without any meaningful error message appearing in stackdriver).

And, using `console.log(JSON.stringify(myStuff))` results in stackdriver crashing with a 500 error: 

```
{
  "error": {
    "code": 500,
    "message": "Failed to convert server response to JSON",
    "status": "INTERNAL"
  }
}
```

**Comment by ernsheong** [comment]
Created an issue at https://issuetracker.google.com/issues/149668307, feel free to chime in

**Comment by techieshark** [comment]
It'd be great if we could just console.*level*() an object with the various stackdriver meta fields set within that logged object -- so we could avoid the headache of async Stackdriver logging in cloud functions.

**Comment by npomfret** [comment]
@techieshark I think the discussion has moved here https://issuetracker.google.com/issues/149668307 - please add your opinion

**Comment by kanekv** [comment]
Will console.* ever log structured objects?

**Comment by npomfret** [comment]
@kanekv I've had real trouble with the console (see https://issuetracker.google.com/issues/149668307).  It fails to log in some situations so I always stringify my data before logging to to console AND prefix that string with something that makes it invalid json.

**Comment by melissammbeff** [comment]
No solution so far. I deployed my pubsub function (nodejs12) and the logs (console.log, console.error) are now not recognizable. When I was using nodejs8 I didn't have this problem, I could filter logs by level. Now they are unstructured logged and it's hard to trace errors.

**Comment by Rothanan** [comment]
Facing the same problem, logging is currently a complete mess.

**Comment by nicoleczhu** [comment]
Hi folks, I’m splitting up this issue since we’re talking about different logging mechanics here: 

1. `console.METHOD defaulting to DEBUG level only`: [Per specs](https://cloud.google.com/functions/docs/monitoring/logging#writing_runtime_logs), it's working as expected: `Most log entries do not have an associated log level. These include logs emitted using console.log(), console.info() , console.warn() or console.error()`

Earlier this year, the functions product team implemented structured logs in cloud functions environment to help users get levels. [Sample here](https://cloud.google.com/functions/docs/monitoring/logging#writing_structured_logs)

If you are still experiencing problems getting severity from structured logs, please indicate your node version and reproducible code snippet to help the us troubleshoot. Since this issue isn't within this repo, we can resume any future [conversation here](https://b.corp.google.com/issues/149668307)

2. `Missing logs issue`. Now tracking here #969. For now we recommend logging via the native runtime log wrappers (approach mentioned in 1)

Let me know if there are other gaps that we can address. Thanks all
